src/stateManager/ValidationService.ts

```diff
@@ -94,16 +96,83 @@ export default class ValidationService {
+        // Objective timestamp validation
+        if (!isValidTimestamp) {
+            if (
+                blk.height <= 0 || // genesis block
+                previousBlock === undefined || // no previous block == genesis block
+                previousBlock.onChainTimestamp !== undefined
+            ) {
+                // already has best timestamp
+                // no point in trying to  update the timestamp => fraud proof
+                const fraudProof =
+                    this.disputeHandler.proofManager.createIncorrectDataProof(
+                        signedBlock
+                    );
+                await this.disputeHandler.createDispute(blk.forkId, "0x00", 0, [
+                    fraudProof
+                ]);
+                return dispute(AgreementFlag.INCORRECT_DATA);
+            }
+
+            // try to fetch later timestamp from on-chain data
+            const onChainTimestamp =
+                await this.fetchOnChainTimestamp(previousBlock);
+
+            // Early return: Couldn't fetch or not better than current
+            if (!onChainTimestamp || onChainTimestamp <= previousTimestamp) {
+                const fraudProof =
+                    this.disputeHandler.proofManager.createIncorrectDataProof(
+                        signedBlock
+                    );
+                await this.disputeHandler.createDispute(blk.forkId, "0x00", 0, [
+                    fraudProof
+                ]);
+                return dispute(AgreementFlag.INCORRECT_DATA);
+            }
+
+            // Update the previous block with the on-chain timestamp
+            previousBlock.onChainTimestamp = onChainTimestamp;
+            this.storage.blocks.setOnChainTimestamp(
+                previousBlock.forkId,
+                previousBlock.height,
+                onChainTimestamp
+            );
 
-        // Check if enough time has passed
-        const timeFlag = await this.isEnoughTimeSubjective(signedBlock, blk);
-        if (timeFlag !== ExecutionFlags.SUCCESS) {
-            return { success: false, flag: timeFlag };
+            // Re-validate with updated timestamp
+            const isValidTimestamp =
+                blk.timestamp >= onChainTimestamp &&
+                blk.timestamp <= onChainTimestamp + this.timeCfg.p2pTime;
+
+            if (!isValidTimestamp) {
+                const fraudProof =
+                    this.disputeHandler.proofManager.createIncorrectDataProof(
+                        signedBlock
+                    );
+                await this.disputeHandler.createDispute(blk.forkId, "0x00", 0, [
+                    fraudProof
+                ]);
+                return dispute(AgreementFlag.INCORRECT_DATA);
+            }
+            // If valid, continue with rest of validation
         }
 
+        // Subjective timestamp validation (requireS)
+        if (
+            Math.abs(Clock.getTimeInSeconds() - blk.timestamp) >=
+            this.timeCfg.agreementTime
+        )
```

Comment:
Commented on Trello - this one should be ignored if the on-chain timestamp is set